ssl-hep/ServiceX_frontend
# feat: Use hatchling as the build backend

* Resolves #344
* Resolves #346

If accepted will supercede other open PRs:

* Closes #345
* Closes #348

## Description

**fix: Use 'isinstance' for type checks**

* Avoids flake8 error: [E721 do not compare types, for exact checks use `is` / `is not`, for instance checks use `isinstance()`](https://www.flake8rules.com/rules/E721.html). This is required to pass linting.

**feat: Use hatchling as the build backend**

The motivation here is that `poetry` solves are quite difficult to debug with and I would say are quite often wrong. @ponyisi had to do quite a bit of work to tune things in `poetry` to get PR #348 passing at all, while just providing lower bounds for the `hatchling` build backend to communicate with the dependency resolvers of `pip` and `uv` resulted in trivial solves once it was understood that the `pydantic` `v2.0` API is not supported.

`poetry` was introduced in d10d1594e11026f106aada83161e88db8aa1e2fd and I can't find any PRs or GitHub Issues where the decision to adopt it was discussed, so there may be reasons that I don't understand for its adoption. However, if the main motivation was to use its lock file feature I would suggest either just using `uv pip compile` (super fast!)

```console
# cold cache
$ time uv pip compile --all-extras --generate-hashes --output-file requirements.lock pyproject.toml
...
real	0m0.942s
user	0m0.243s
sys	0m0.319s
# warm cache
$ time uv pip compile --all-extras --generate-hashes --output-file requirements.lock pyproject.toml
...
real	0m0.127s
user	0m0.082s
sys	0m0.106s
```

or if you want a fully integrated system using `pdm`.

* Switch to use hatchling as the build backend from poetry.
   - Provide lower bounds for all dependencies. For dependencies that are used but are dependencies of other dependencies, and so have their versions controlled by them, list without a lower bound.
   - Add upper bound on pydantic dependency as the v2.X API change is known to be fully breaking.
   - Add missing databinder dependencies of rich and nest-asyncio.
   - Remove the jupyter metapackage as a dependency as it obscures the actual dependencies and is not a runtime requirement.
   - Support Python 3.8+.
* Add 'develop' extra that includes the 'pandas', 'test', and 'docs' extras.
* Add PyPI metadata to pyproject.toml.
* Remove poetry.lock.

**ci: Add Python 3.12 testing to CI**

* Use pip to install dependencies.
* Only install the required 'servicex[pandas,test]' dependencies to test that the install requirements are correct.
* Remove install of caio from sdist, as no longer required.

**ci: Use PyPA tooling for publishing**

* As hatchling has been adopted as the build backend, removing poetry, the publishing workflow needs to be updated to use the standard PyPA tooling of build and twine.
   - PyPI's Trusted Publisher technology should be used, but that's a seperate PR.
* Use sed to update the version metadata in pyproject.toml.
   - Search for 'version =' in pyproject.toml and then replace the contents on that line in between double quotes with the version set.
* List the contents of the sdist and wheel for visual checks in the logs.
* Upload the distributions to GitHub for recovery or debugging if needed.
* Use sphinx to build the docs.
